jep-007-filter-expressions.md

```diff
@@ -174,7 +174,7 @@ Using the previous example, given the following data:
          {"state": "CA", "value": 4}]}
 ```
 
-The expression `foo[?state == \`WA\`]` will return the following value:
+The expression ``foo[?state == `WA`]`` will return the following value:
```

Comment:
Many of the early JEPs contain examples like this, which use the now-invalid syntax. They should probably be updated to use JEP-12 syntax (e.g., ``` `WA` ``` → `'WA'` here), but also in a followup.

Reply:
I would tend to be OK with this syntax in pre JEP-12 proposals as it was legal back then.

Reply:
Well, what is the definition of "pre JEP-12"? The legacy behavior seems to have been introduced in [JEP-7](https://github.com/jmespath-community/jmespath.spec/blob/HEAD/jep-007-filter-expressions.md#literal-expressions) ("_an alternate syntax is supported where the starting and ending double quotes are not required for strings_"), but it appears in the earlier [JEP-3](https://github.com/jmespath-community/jmespath.spec/blob/HEAD/jep-003-functions.md) and [JEP-6](https://github.com/jmespath-community/jmespath.spec/blob/HEAD/jep-006-improved-identifiers.md#rationale) and also in [JEP-11](https://github.com/jmespath-community/jmespath.spec/blob/HEAD/jep-011-let-function.md) (which is numerically lower than JEP-12, but was accepted after it).

It seems like we are in a state where reading those JEPs in the future will be unnecessarily challenging. Perhaps the best fix is updating all three to use ``` `"…"` ``` like [JEP-17](https://github.com/jmespath-community/jmespath.spec/blob/main/jep-017-root-reference.md#motivation), with an explanatory forward reference to JEP-7 in the former two?

Reply:
Your digging for when legacy syntax was introduced made me look in more details.

It appears that JEPs have not been designed – or at least, accepted – in chronological order.
For instance, while you rightly point out that legacy syntax [was introduced with JEP-7, accepted in 16-Dec-2013](https://github.com/jmespath/jmespath.py/commit/95a393e32e5aeb5f984cf2d7811fad75d4e0df95). it appears that JEP-3 – where many examples use legacy syntax – was [authored retroactively](https://github.com/jmespath/jmespath.py/commit/808090197fa6b25f01fd0cf9a9a0b0c3bf07e36f) after JEP-7.

That partly explains why legacy syntax appears in _earlier_ JEPs.

As far as this project is concerned, I agree that we should strive for consistency but we can not rewrite history and iron out all mistakes that were made. For instance, we would face a dilemma:

- We can not use the `` `"foo"` `` syntax in JEP-3 as JSON `literal` expressions were not supposed to even exist at that time.
- We can not use the `` 'foo' `` syntax in JEP-3 as `raw-string` was only introduced later with JEP-12.

Reply:
> We can not use the `` `"foo"` `` syntax in JEP-3 as JSON `literal` expressions introduced with JEP-7 were not supposed to even exist at that time.

Right, that's why I'm also suggesting explanatory forward reference to JEP-7. For example:

**JEP-3**
`````patch
@@ -171,6 +171,9 @@ return_type foo(array[number] $argname)
 As a shorthand, the type `any` is used to indicate that the argument can be
 of any type (`array|object|number|string|boolean|null`).
 
+The below examples make use of [JEP-7 literal expressions](./jep-007-filter-expressions.md#literal-expressions)
+such as strings like `"abc"` and numbers like `-1`.
+
 The first function below, `abs` is discussed in detail to demonstrate the
 above points.  Subsequent function definitions will not include these details
 for brevity, but the same rules apply.
@@ -276,7 +279,7 @@
 |---|---
 | `abs(1)` |1
 | `abs(-1)`|1
-| ``abs(`abc`)``| |`<error: invalid-type>`
+| ``abs(`"abc"`)``| |`<error: invalid-type>`
 
 ### avg
 
@@ … @@
`````

**JEP-6**
`````patch
@@ -157,8 +157,8 @@ escaped-char      = escape (
 Adopting the same string rules as JSON strings will allow users familiar with
 JSON semantics to understand how JMESPath identifiers will work.
 
-This change also provides a nice consistency for the literal syntax proposed
-in JEP 3.  With this model, the supported literal strings can be the same
+This change also provides a nice consistency with [JEP-7 literal expressions](./jep-007-filter-expressions.md#literal-expressions).
+With this model, the supported literal strings can be the same
 as quoted identifiers.
 
 This also will allow the grammar to grow in a consistent way if JMESPath
@@ -172,8 +172,8 @@ this is just a suggested syntax, not a formal proposal), given the data:
 You can now have the following JMESPath expressions:
 
 ```
-foo[?"✓" = `✓`]
-foo[?"\u2713" = `\u2713`]
+foo[?"✓" = `"✓"`]
+foo[?"\u2713" = `"\u2713"`]
 ```
 
 As a general property, any supported JSON string is now a supported quoted
`````

**JEP-11**
`````patch
@@ -50,7 +50,7 @@
 In this example we can hard code the state `WA`:
 
 ```
-states[?name==`WA`].cities[]
+states[?name==`"WA"`].cities[]
 ```
 
 but it is not possible to base this on a value of `first_choice`, which
@@ … @@
`````